### ISSUE TITLE
Plane Restrictions per Player

### ISSUE THREAD

**Issue by kevATin** [biell/alti-server#23]
Is it possible to define "/add plane xyz" restrictions/permissions per player instead of per server/map?

**Comment by biell** [comment]
I don't have anything coded, but Xal does have an interface.  See `allowPlayerPerk` in https://gitlab.com/alti2d/server_patches

**Comment by kevATin** [comment]
So what I should do is disallow every single perk, and then allow the perks the player chooses for each individual player. Sounds simple enough, thx~

Calling those commands should work fine in the scripts when xx is enabled, right?

Offtopic: What are teams 0, 1 and 2 used for?

**Comment by biell** [comment]
Yes, you can call all those commands in a script with server::send.  I would think about writing a function outside of the HOOKS { ... } definition to do the heavy lifting, so your hooks are cleaner.  I think those restrictions are for as long as the player is connected, so you could set it when the player joins.

Team 0 is left, 1 is right, and 2 is spec.
Teams 2, 3, 4, ... are numbers assigned to the colors, starting with red and ending with brown.

**Comment by kevATin** [comment]
Yeah I was thinking about using functions, will do.

Is there any way I could get the alti server to remember the state of players and maps and all? Even over restarts? Or will I have to write the state of everything to a file and then read that file at every restart, when players join, etc?

According to xal's site -1 is used for spec, so are both -1 and 2 just specs or somehow different?

**Comment by biell** [comment]
-1 is spec under certain conditions (like `assignTeam`) and 2 is spec under other conditions (e.g. team number).  The everything was thought out as well as it could have been.

I have functions to help you out for saving state, but not across restarts of the server.  You should use a database for that.

If you say  `mem->player('coins', 'backpack', $p)+=10;`, I will create an object called `coins.backpack.$p`, where `$p` is translated to the player's vaporID, if it doesn't exist already, and add `10` to it.  I will remove these objects after I haven't seen the player in a day.

If you use `mem->game(...)`, I will also store things the same way, but I will clean those memory objects up after every map change.

I have a `mem->server(...)` storage mechanism also, but I will not save these values across reboots. The values will stay around for the lifetime of the server.

I could create a `mem->persist(...)` interface (or modify `mem->server(...)` to persist, and store the data elements in the database.  But, you would have to keep it clean yourself, otherwise it would just grow forever.

My mem->?? objects are valid in Lvalue and Rvalue contexts.

**Comment by kevATin** [comment]
Overwriting existing ```mem->server(...)``` behaviour doesn't sound like a good idea, maybe someone needs it to automatically flush on server restart~

A separate ```mem->persist(...)``` sounds great. It could be used for things that don't need to be accessed often enough to require to be put into a dedicated database. And for testing purposes with low player count it would also be useful as a substitute for the database altogether.

> But, you would have to keep it clean yourself, otherwise it would just grow forever.

What would be the best way to check for the last time a player was online on the server?
